lib/SILOptimizer/LoopTransforms/LICM.cpp

```diff
@@ -963,6 +963,41 @@ static bool storesCommonlyDominateLoopExits(SILValue addr, SILLoop *loop,
   if (stores.count(header) != 0)
     return true;
 
+  // Also a store in the pre-header dominates all exists. Although the situation
+  // is a bit different here: the store in the pre-header remains - it's not
+  // (re)moved by the LICM transformation.
+  // But even if the loop-stores are not dominating the loop exits, it
+  // makes sense to move them out of the loop if this case. When this is done,
+  // dead-store-elimination can then most likely eliminate the store in the
+  // pre-header.
+  //
+  //   pre_header:
+  //     store %v1 to %addr
+  //   header:
+  //     cond_br %cond, then, tail
+  //   then:
+  //     store %v2 to %addr    // a conditional store in the loop
+  //     br tail
+  //   tail:
+  //     cond_br %loop_cond, header, exit
+  //   exit:
+  //
+  //  will be transformed to
+  //
+  //   pre_header:
+  //     store %v1 to %addr    // <- can be removed by DSE afterwards
+  //   header:
+  //     cond_br %cond, then, tail
+  //   then:
+  //     br tail
+  //   tail:
+  //     cond_br %loop_cond, header, exit
+  //   exit:
+  //     store %v2 to %addr
```

Comment:
If %cond was never true, then this is storing the wrong value ??

Reply:
I should use a different name for %v2 here. It's actually the value which comes out of the SSAUpdater.

Reply:
yeah, I would hope it's inserting a phi inside the loop

Reply:
Would it be worth writing a unit test for the example from the documentation?

Reply:
The test I added is exactly this. I added a check for the phi argument.